Documentation/bpf.rst

```diff
@@ -2964,13 +2964,266 @@ Program Types
+  XDP's early processing stage, this effectively 'pretends' to the kernel's
+  networking stack that these packets have never been seen by the networking
+  device. Additionally, if a potential bug in the stack's receive path
+  got uncovered and would cause a 'ping of death' like scenario, XDP can be
+  utilized to drop such packets right away without having to reboot the
+  kernel or restart any services, meaning, XDP can just atomically swap
+  such drop enforcement of bad packets without any network traffic
+  interruption on an end host.
+
+  Another use case for pre-stack processing is that given the kernel has not
+  yet allocated an ``skb`` for the packet, the BPF program is free to modify
+  the packet and, again, have it 'pretend' to the stack that it was received
+  by the networking device this way. This allows for cases such as having
+  custom packet mangling and encapsulation protocols where the packet can be
+  decapsulated prior to entering GRO aggregation in which GRO otherwise would
+  not be able to perform any sort of aggregation due to not being aware of
+  the custom protocol. XDP also allows to push metadata (non-packet data) in
+  front of the packet. This is 'invisible' to the normal kernel stack, can
+  be GRO aggregated (for matching metadata) and later on processed in
+  coordination with a tc ingress BPF program where it has the context of
+  a ``skb`` available for e.g. setting various skb fields.
+
+..
+
+* **Flow sampling, monitoring**
+
+  XDP can also be used for cases such as packet monitoring, sampling or any
+  other network analytics, for example, as part of an intermediate node in
+  the path or on end hosts in combination also with prior mentioned use cases.
+  For complex packet analysis, XDP provides a facility to efficiently push
+  network packets (truncated or with full payload) and custom metadata into
+  a fast lockless per CPU memory mapped ring buffer provided from the Linux
+  perf infrastructure to an user space application. This also allows for
+  cases where only a flow's initial data can be analyzed and once determined
+  as good traffic having the monitoring bypassed. Thanks to the flexibility
+  brought by BPF, this allows for implementing any sort of custom monitoring
+  or sampling.
+
+..
+
+**XDP operation modes**
+
+* **Native XDP**
+
+  TODO
+
+..
+
+* **Offloaded XDP**
+
+  TODO
+
+..
+
+* **Generic XDP**
+
+  TODO
+
+..
+
+**Driver support**
+
+Since BPF and XDP is evolving quickly in terms of feature and driver support,
+the following lists native and offloaded XDP drivers at the time of this
+writing.
+
+**Drivers supporting native XDP**
+
+* **Broadcom**
+
+  * bnxt
+
+..
+
+* **Cavium**
+
+  * thunderx
+
+..
+
+* **Intel**
+
+  * ixgbe
+  * ixgbevf
+  * i40e
+
+..
+
+* **Mellanox**
+
+  * mlx4
+  * mlx5
+
+..
+
+* **Netronome**
+
+  * nfp
+
+..
+
+* **Solarflare**
+
+  * sfc (currently out of tree driver, but upstreamed soon)
```

Comment:
Might be useful to actually put a version marker here to say "out of tree driver as at Linux v4.16, but is expected to become upstream soon".

Equally it might be kinda nice to have kernel version support numbers for other drivers, but if that's too much hassle you could put a short paragraph at the start of this drivers section that describes that this section relates to Linux v4.16.

Reply:
If we go through to figure out all the driver support in different Linux versions (if they vary), it might be easier to represent this all as a table.

Reply:
Agreed, this could be an additional table. I'll state that this is the current state as of v4.17. I like that also in dpdk doc, it's a simple and easy to grasp list (https://dpdk.org/doc/nics). A feature matrix should also include other features like return codes supported by the NICs that would be relevant to XDP.